sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1111 @@
+
+        for n in range(self.n_iter):
+            train = []
+            test = []
+
+            for i, cls in enumerate(classes):
+                permutation = rng.permutation(cls_count[i])
+                cls_i = np.where((y == cls))[0][permutation]
+
+                train.extend(cls_i[:n_i[i]])
+                test.extend(cls_i[n_i[i]:n_i[i] + t_i[i]])
+
+            # Because of rounding issues (as n_train and n_test are not
+            # dividers of the number of elements per class), we may end
+            # up here with less samples in train and test than asked for.
+            if len(train) < n_train or len(test) < n_test:
+                # We complete by affecting randomly the missing indexes
+                missing_idx = np.where(bincount(train + test,
+                                                minlength=len(y)) == 0)[0]
+                missing_idx = rng.permutation(missing_idx)
+                train.extend(missing_idx[:(n_train - len(train))])
+                test.extend(missing_idx[-(n_test - len(test)):])
+
+            train = rng.permutation(train)
+            test = rng.permutation(test)
+
+            yield train, test
+
+    def n_splits(self, X=None, y=None, labels=None):
+        return self.n_iter
+
+
+class PredefinedSplit(_PartitionIterator):
+    """Predefined split cross validation iterator
+
+    Splits the data into training/test set folds according to a predefined
+    scheme. Each sample can be assigned to at most one test set fold, as
+    specified by the user through the ``test_fold`` parameter.
+
+    Read more in the :ref:`User Guide <split>`.
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import PredefinedSplit
+    >>> X = np.array([[1, 2], [3, 4], [1, 2], [3, 4]])
+    >>> y = np.array([0, 0, 1, 1])
+    >>> test_folds = [0, 1, -1, 1]
+    >>> ps = PredefinedSplit()
+    >>> ps.n_splits(X, y, labels=test_folds)
+    2
+    >>> print(ps)       # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    PredefinedSplit()
+    >>> for train_index, test_index in ps.split(X, y, labels=test_folds):
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [1 2 3] TEST: [0]
+    TRAIN: [0 2] TEST: [1 3]
+    """
+
+    def _iter_test_indices(self, X, y=None, labels=None):
+        test_fold, unique_folds = self._get_test_folds(X, y, labels)
+        for f in unique_folds:
+            yield np.where(test_fold == f)[0]
+
+    def _get_test_folds(self, X, y, labels):
+        test_fold = column_or_1d(np.array(labels, dtype=np.int))
+        unique_folds = np.unique(test_fold)
+        return test_fold, unique_folds[unique_folds != -1]
+
+    def n_splits(self, X, y=None, labels=None):
+        return len(self._get_test_folds(X, y, labels)[1])
+
+
+class CVIterator(_PartitionIterator):
+    """Wrapper class for old style cv objects and iterables."""
```

Comment:
The reason why I feel this could better be a private class is that, user won't be needing this for themselves since this is mainly for wrapping old cv to make it compatible...

But since it also wraps iterables, I am not entirely sure if we should keep it private... Please advise!